include/dali/core/mm/detail/free_list.h

```diff
@@ -0,0 +1,327 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_MM_DETAIL_FREE_LIST_H_
+#define DALI_CORE_MM_DETAIL_FREE_LIST_H_
+
+#include <cassert>
+#include <utility>
+#include "dali/core/util.h"
+#include "dali/core/mm/detail/align.h"
+
+namespace dali {
+namespace mm {
+
+namespace free_list_impl {
+
+// Deletes a unidirectional list
+template <typename Block>
+void delete_list(Block *&b) {
+  while (b) {
+    Block *n = b->next;
+    delete b;
+    b = n;
+  }
+}
+
+}  // namespace free_list_impl
+
+/**
+ * @brief Maintains a list of free memory blocks of uniform size.
+ *
+ * This object is used for managing a list of perfectly interchangeable memory regions.
+ * As such, it does not need to store size of the blocks, since they will, by definition,
+ * have equal size and the list doesn't need to know what that size is.
+ *
+ * This is a non-intrusive free list and, as such, it needs to allocate the storage for the block
+ * descriptors. The block descriptors are not freed until the object is destroyed.
+ *
+ * If a block is taken from the free list, it's removed from the main list and stored in an
+ * auxiliary list of unused block descriptors. These blocks will be reused when something
+ * is added to the list to avoid interaction with the heap.
+ */
+class uniform_free_list {
+ public:
+  struct block {
+    block *next;
+    void *mem;
+  };
+
+  uniform_free_list() = default;
+  uniform_free_list(const uniform_free_list &) = delete;
+  uniform_free_list(uniform_free_list &&other) {
+    swap(other);
+  }
+
+  void swap(uniform_free_list &other) {
+    std::swap(head_, other.head_);
+    std::swap(unused_blocks_, other.unused_blocks_);
+  }
+
+  uniform_free_list &operator=(uniform_free_list &&other) {
+    if (this != &other) {
+      clear();
+      swap(other);
+    }
+    return *this;
+  }
+
+  ~uniform_free_list() {
+    clear();
+  }
+
+  void clear() {
+    free_list_impl::delete_list(head_);
+    free_list_impl::delete_list(unused_blocks_);
+  }
+
+  void *get() {
+    if (block *blk = head_) {
+      void *mem = blk->mem;
+      head_ = blk->next;
+      blk->next = unused_blocks_;
+      blk->mem  = nullptr;
```

Comment:
I would move this to L92 to keep operation on `blk->mem`. But this is rather a nitpick.